File: samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/sessions/CustomerSession.java
package systems.zlink.samples.deliverydispatch.server.customergateway.sessions;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher;
import systems.zlink.framework.streams.ZLinkStreamError;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class CustomerSession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers;

    public CustomerSession(
            ZLinkSessionContext context,
            ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers) {
        this.context = context;
        this.handlers = handlers;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        return CompletableFuture.allOf(
                context.actors().bound().stream()
                        .map(actor -> actor.notifyDisconnected().toCompletableFuture())
                        .toArray(CompletableFuture[]::new));
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        return handlers.tryHandle(context, dispatch, payload)
                .thenCompose(
                        handled -> {
                            if (handled) return CompletableFuture.completedFuture(null);
                            ZLinkSessionActor actor =
                                    switch (context.actors().bound().size()) {
                                        case 1 -> context.actors().bound().get(0);
                                        case 0 ->
                                                throw new IllegalStateException(
                                                        "Client must subscribe before relaying"
                                                                + " packet '"
                                                                + dispatch.packetName()
                                                                + "'");
                                        default ->
                                                throw new IllegalStateException(
                                                        "Exactly one customer actor must be bound"
                                                                + " before relaying packet '"
                                                                + dispatch.packetName()
                                                                + "'");
                                    };
                            return actor.relay(payload).thenApply(ignored -> null);
                        });
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/sessions/CustomerSession.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.sessions

import systems.zlink.framework.kotlin.ZLinkSuspendingSession
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher
import systems.zlink.framework.streams.ZLinkStreamError

class CustomerSession(
    private val sessionContext: ZLinkSessionContext,
    private val handlers: ZLinkSessionPacketDispatcher<ZLinkSessionContext>,
) : ZLinkSuspendingSession() {
    override fun context(): ZLinkSessionContext = sessionContext

    override suspend fun onConnectedSuspending() {}

    override suspend fun onDisconnectedSuspending() {
        for (actor in sessionContext.actors().bound()) actor.notifyDisconnected().await()
    }

    override suspend fun onErrorSuspending(error: ZLinkStreamError) {}

    override suspend fun onDispatchSuspending(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        val handled = handlers.tryHandle(sessionContext, dispatch, payload).await()
        if (handled) {
            return
        }
        val actor =
            when (sessionContext.actors().bound().size) {
                1 -> sessionContext.actors().bound()[0]
                0 ->
                    error("Client must subscribe before relaying packet '${dispatch.packetName()}'")
                else ->
                    error(
                        "Exactly one customer actor must be bound before relaying packet '${dispatch.packetName()}'"
                    )
            }
        actor.relay(dispatch, payload).await()
    }
}
